model/inline/src/main/scala/aqua/model/inline/tag/StreamRestrictions.scala

```diff
@@ -0,0 +1,29 @@
+package aqua.model.inline.tag
+
+import aqua.model.inline.state.{Arrows, Config, Exports, Mangler}
+import aqua.model.{OpModel, RestrictionModel}
+import aqua.types.StreamType
+
+import cats.data.{Chain, State}
+import cats.syntax.traverse.*
+
+object StreamRestrictions {
+
+  // restrict streams that are generated in a tree
+  def restrictStreams[S: Mangler: Exports: Arrows: Config](
+    childrenToModel: Chain[OpModel.Tree] => OpModel.Tree
+  )(children: Chain[State[S, OpModel.Tree]]): State[S, OpModel.Tree] =
+    children.traverse(restrictStreamsAround).map(childrenToModel)
```

Comment:
Is this function really needed? Its signature is three time bigger than the body. I think it will be even more readable if inliners have this code directly.